pkg/authorization/admission/restrictnodes/restrictnodes.go

```diff
@@ -0,0 +1,222 @@
+		kclient: kclient,
+	}, nil
+}
+
+var (
+	secretsResource = kapi.Resource("secrets")
+	nodesResource   = kapi.Resource("nodes")
+	podsResource    = kapi.Resource("pods")
+)
+
+func (q *restrictNodesAdmission) Admit(a admission.Attributes) error {
+	// We only care about nodes we can identify
+	userInfo := a.GetUserInfo()
+	if userInfo == nil {
+		return nil
+	}
+	username := userInfo.GetName()
+	if !strings.HasPrefix(username, "system:node:") {
+		return nil
+	}
+	nodeName := strings.TrimPrefix(username, "system:node:")
+
+	// We only care about secrets, pods, and nodes; ignore anything else.
+	switch a.GetResource().GroupResource() {
+	case secretsResource:
+		return q.admitSecret(nodeName, a)
+	case nodesResource:
+		return q.admitNode(nodeName, a)
+	case podsResource:
+		return q.admitPod(nodeName, a)
+	default:
+		return nil
+	}
+}
+
+var (
+	everything = labels.Everything()
+)
+
+func (q *restrictNodesAdmission) admitSecret(nodeName string, a admission.Attributes) error {
+	ns := a.GetNamespace()
+	name := a.GetName()
+	pods, err := q.pods.Pods(ns).List(everything)
+	if err != nil {
+		return err
+	}
+	for _, pod := range pods {
+		if pod.Spec.NodeName != nodeName {
+			continue
+		}
+		found := false
+		visitSecretNames(pod, func(referencedName string) bool {
+			if name == referencedName {
+				found = true
+				return false
+			}
+			return true
+		})
+		if found {
+			return nil
+		}
+	}
+
+	// TODO: other secret references nodes need? PVs?
+
+	return kapierrors.NewForbidden(a.GetResource().GroupResource(), a.GetName(), errors.New("cannot access secret"))
+}
+
+func visitSecretNames(pod *kapi.Pod, visitor func(string) bool) {
+	for _, s := range pod.Spec.ImagePullSecrets {
+		if !visitor(s.Name) {
+			return
+		}
+	}
+	for _, v := range pod.Spec.Volumes {
+		if v.Secret != nil && !visitor(v.Secret.SecretName) {
+			return
+		}
+	}
+	for _, v := range pod.Spec.InitContainers {
+		for _, e := range v.Env {
+			if e.ValueFrom != nil && e.ValueFrom.SecretKeyRef != nil && !visitor(e.ValueFrom.SecretKeyRef.Name) {
+				return
+			}
+		}
+	}
+	for _, v := range pod.Spec.Containers {
+		for _, e := range v.Env {
+			if e.ValueFrom != nil && e.ValueFrom.SecretKeyRef != nil && !visitor(e.ValueFrom.SecretKeyRef.Name) {
+				return
+			}
+		}
+	}
+}
+
+func (q *restrictNodesAdmission) admitNode(nodeName string, a admission.Attributes) error {
+	// Only let nodes modify their own API resource
+	if a.GetName() == nodeName {
+		return nil
+	}
+	return kapierrors.NewForbidden(a.GetResource().GroupResource(), a.GetName(), errors.New("cannot access other nodes"))
+}
+
+func (q *restrictNodesAdmission) admitPod(nodeName string, a admission.Attributes) error {
+	if len(a.GetSubresource()) > 0 && a.GetSubresource() != "status" {
+		return nil
+	}
+
+	switch a.GetOperation() {
+	case admission.Create:
+		// only let nodes create mirror pods targeted to themselves
+		object := a.GetObject()
+		if object == nil {
+			// TODO: fail closed?
```

Comment:
Could have the downside of blowing up the cluster... but i'll take a little security